### ISSUE TITLE
'WaitForExternalEvent' does not respect explicit infinite timeout

### ISSUE THREAD

**Issue by julealgon** [Azure/azure-functions-durable-extension#2111]
### Description

I have a custom "TTL tracking function" that will fire when a resource is created, wait until the resource's specified "KeepUntil" field is reached, and then fire a deletion message to remove the resource. If the resource is deleted manually before the "KeepUntil" is reached, the orchestration is notified and stops tracking the resource.

This relies on a timer, as well as an external event that is triggered when someone manually deletes the resource.

However, `WaitForExternalEvent` doesn't seem to respect `Timeout.Infinite`, and instead handles it as any other TimeSpan value and abruptly ends with a `TimeoutException`.

I understand there is already another overload for infinite waits but it doesn't allow for cancellation. I wanted to be able to pass the cancellation token to cancel the task that "lost" the race regardless of which one it was.

### Expected behavior
`Timeout.Infinite` should be handled in the `WaitForExternalEvent` implementation and result in an infinite wait instead of an immediate `TimeoutException`.

Bonus: Other negative values for the timeout parameter should throw `ArgumentOutOfRangeException`.

### Actual behavior
The call immediately throws `TimeoutException` as if "the timeout has passed" (`-1` miliseconds have passed).

### Relevant source code snippets
Full logic:
```csharp
        using var cancellationTokenSource = new CancellationTokenSource();
        var expirationTimeTask = context.CreateTimer(parameter.KeepUntil, cancellationTokenSource.Token);
        var stopTrackingEventTask = context.WaitForExternalEvent("StopTracking", Timeout.InfiniteTimeSpan, cancellationTokenSource.Token);

        var winner = await Task.WhenAny(expirationTimeTask, stopTrackingEventTask);

        // Cancel the loser task
        cancellationTokenSource.Cancel();
        if (winner == expirationTimeTask)
        {
            // call to delete object
        }
```

Offending line:
```csharp
        var stopTrackingEventTask = context.WaitForExternalEvent("StopTracking", Timeout.InfiniteTimeSpan, cancellationTokenSource.Token);
```

### Known workarounds
I'll have to use the overload that doesn't take a cancellation token for now.

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**: 2.6.1
- **Azure Functions runtime version (1.0 or 2.0)**: 4.0
- **Programming language used**: C#

**Comment by cgillum** [comment]
Agreed that we will need to add special handling for `Timeout.InfiniteTimeSpan`. I'm marking this issue as a bug and putting it in the backlog. The fix will need to go somewhere around [this line of code](https://github.com/Azure/azure-functions-durable-extension/blob/5a4251d3900ec1fcbca03267c4aa370a25189aa5/src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs#L974).
